### ISSUE TITLE
[BUG] Mapstruct mappers are wrong if lombok needs to create Builders in later rounds

### ISSUE THREAD

**Issue by fraenkelc** [projectlombok/lombok#3116]
**Describe the bug**
the lombok implementation of `AstModifyingAnnotationProcessor` in https://github.com/projectlombok/lombok/blob/560ded50127b12bd8168f8ae2bada47253f15bc8/src/bindings/mapstruct/lombok/mapstruct/NotifierHider.java#L19-L33 does not work correctly when lombok gets called in a subsequent round to modify a file emitted by a different AP in an earlier round (that contains `@SuperBuilder` or `@Builder`). 

Instead of reporting to mapstruct if a given type is "complete" lombok reports "complete" for all files as long as it has run once.

**To Reproduce**
I've uploaded a reproducer project at https://github.com/fraenkelc/bug-lombok-mapstruct-ap

Steps:
1. create an AP that writes a source file which contains the `@SuperBuilder` (probably also affects `@Builder`) lombok annotation
2. Use this AP together with Lombok, mapstruct and the lombok-mapstruct binding in a project
3. Write a Mapper that maps **to** the generated file
4. compile
5. Mapstruct will run too early and not see the modified ast and in turn not use the lomboks builder
  https://github.com/fraenkelc/bug-lombok-mapstruct-ap/blob/aa68ee421703115af6cd7c19d1045d3e110d506c/README.md?plain=1#L21-L43

**Expected behavior**
Lombok should correctly delay Mapstruct until it is done, e.g. by implementing AstModifyingAnnotationProcessor with a per-type logic instead of the current "ran once and done" logic.

**Version info (please complete the following information):**
 - 1.18.22
 - Java 11.0.14.1 (Eclipse Adoptium)

**Comment by Drevsh** [comment]
I think I'm running into that issue as well.

I have a DTO class which uses `@Builder`. When I try to map to this DTO I get the following error message from mapstrcut:

> Unknown property "addOns" in result type Item.ItemBuilder. Did you mean "null"?

Interesting enough I only get the issue with Collections but everything else works fine.

The DTO looks as follows:


```
@Builder
@Getter
@Setter 
@NoArgsConstructor  
@AllArgsConstructor  
public class Item {  
	private String comment;  
	private Collection<Long> addOns;  
}
```

Without the addOns collection mapstruct uses the builder, but ignores the collection.
```
ItemBuilder item1 = Item.builder();
item1.comment( item.getComment() );
// some more mappings, but addOns is missing

```
As soon as I  also want the collection mapped I get an error.

**Comment by Rawi01** [comment]
@Drevsh Do you use an additional annotation processor to generate a new class with lombok annotations and try to map this one using mapstruct? If not your problem is something different. You can try if it is a lombok problem by replacing the lombok version with a delomboked version of your class.

@fraenkelc Thanks for providing the example project. I tried to fix this by adding all processed classes to a `Set` and let the binding check if it contains the `TypeMirror`. This doesn't work as expected because MapStruct also calls this method for unprocessed types e.g. `java.util.ArrayList`. I think we can somehow solve this but it might take some time.

**Comment by oenie** [comment]
@Rawi01 Have you tried (if that's possible) to turn off the useBuilder feature in your mapstruct ?

@Mapper(componentModel = "cdi", **builder = @Builder(disableBuilder = true)**)

**Comment by jakabgyonci97** [comment]
> I think I'm running into that issue as well.
> 
> I have a DTO class which uses `@Builder`. When I try to map to this DTO I get the following error message from mapstrcut:
> 
> > Unknown property "addOns" in result type Item.ItemBuilder. Did you mean "null"?
> 
> Interesting enough I only get the issue with Collections but everything else works fine.
> 
> The DTO looks as follows:
> 
> ```
> @Builder
> @Getter
> @Setter 
> @NoArgsConstructor  
> @AllArgsConstructor  
> public class Item {  
> 	private String comment;  
> 	private Collection<Long> addOns;  
> }
> ```
> 
> Without the addOns collection mapstruct uses the builder, but ignores the collection.
> 
> ```
> ItemBuilder item1 = Item.builder();
> item1.comment( item.getComment() );
> // some more mappings, but addOns is missing
> ```
> 
> As soon as I also want the collection mapped I get an error.

Strange enough, I had a similar issue. In my case the addOn property was a simple String. Would it be possible that the problem is the property name? Like addOn being a specifal word?

**Comment by rzwitserloot** [comment]
Gracias to @kanchev1 for fixing this one! Will be in next lombok release.

**Comment by rzwitserloot** [comment]
... oof, have to pull the rug on this one. @kanchev1's fix is getting reverted now and is as far as we can tell fundamentally the wrong approach. Comments as to why kanchev's approach appears to be a dead end will be posted as comments on the PR, which is #3744 .

**Comment by savinov** [comment]
The issue confirmed:
- lombok 1.18.30
- mapstruct 1.6.3
- DTO with SuperBuilder annotation
- use DTO field in Mapping target
- DTO with filed of any type named **addSomething** leads to Unknown property "addSomething"
- DTO with filed of any type named **adzSomething** leads to successful compilation

> > I think I'm running into that issue as well.
> > I have a DTO class which uses `@Builder`. When I try to map to this DTO I get the following error message from mapstrcut:
> > > Unknown property "addOns" in result type Item.ItemBuilder. Did you mean "null"?
> > 
> > 
> > Interesting enough I only get the issue with Collections but everything else works fine.
> > The DTO looks as follows:
> > ```
> > @Builder
> > @Getter
> > @Setter 
> > @NoArgsConstructor  
> > @AllArgsConstructor  
> > public class Item {  
> > 	private String comment;  
> > 	private Collection<Long> addOns;  
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Without the addOns collection mapstruct uses the builder, but ignores the collection.
> > ```
> > ItemBuilder item1 = Item.builder();
> > item1.comment( item.getComment() );
> > // some more mappings, but addOns is missing
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > As soon as I also want the collection mapped I get an error.
> 
> Strange enough, I had a similar issue. In my case the addOn property was a simple String. Would it be possible that the problem is the property name? Like addOn being a special word?

**Comment by yvasyliev** [comment]
Hi @rzwitserloot why was this issue closed if the #3744 was reverted? I still see a problem with Lombok/MapStruct integration -https://github.com/mapstruct/mapstruct/issues/3962
